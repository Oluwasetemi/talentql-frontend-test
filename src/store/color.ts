// create the color slice with it reducers
// checkColor
// unCheckColor
// restoreColors
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import type { IColors } from '../App';
import { colors } from '../filter.json';

export type colorState = {
	colors: Array<IColors>;
	selectedColors: Array<string>;
};

const colorTypes = colors.map((color) => color.color);

export const initialState: colorState = {
	colors,
	selectedColors: [...colorTypes],
};

// function isChecked(selectedColors: string[], color: string) {
// 	return selectedColors.includes(color);
// }

export const colorSlice = createSlice({
	name: 'Colors',
	initialState,
	reducers: {
		isChecked(colors, action) {
			colors.selectedColors.includes(action.payload);
		},
		checkColor(colors, action: PayloadAction<string>) {
			colors.selectedColors.push(action.payload);
		},
		unCheckColor(colors, action) {
			// console.log(action);
			const index = colors.selectedColors.findIndex(
				(color) => color === action.payload,
			);
			colors.selectedColors.splice(index, 1);
			// console.log(a);
		},
		resetColors(colors) {
			colors.selectedColors = colorTypes;
		},
	},
});

export const { isChecked, checkColor, unCheckColor, resetColors } =
	colorSlice.actions;
export default colorSlice.reducer;
