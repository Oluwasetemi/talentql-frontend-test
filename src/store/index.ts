import { configureStore } from '@reduxjs/toolkit';
import colorReducer from './color';
import shapeReducer from './shape';

const store = configureStore({
	reducer: {
		color: colorReducer,
		shape: shapeReducer,
	},
});
export default store;

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;
