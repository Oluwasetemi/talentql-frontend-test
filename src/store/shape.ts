// create the color slice with it reducers
// checkColor
// unCheckColor
// restoreColors
import { createSlice } from '@reduxjs/toolkit';
import { shapes } from '../filter.json';

export type shapeState = {
	shapes: typeof shapes;
	selectedShapes: typeof shapes;
};
const initialState: shapeState = {
	shapes,
	selectedShapes: shapes,
};

export const shapeSlice = createSlice({
	name: 'Shapes',
	initialState,
	reducers: {
		checkShape: (shapes, action) => {},
		unCheckShape: (shapes, action) => {},
		restoreShapes: (shapes, action) => {},
	},
});

export const { checkShape, unCheckShape, restoreShapes } = shapeSlice.actions;
export default shapeSlice.reducer;
