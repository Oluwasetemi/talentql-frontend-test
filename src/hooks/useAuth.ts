import React from 'react';
import { authContext } from '../context/authContext';

export function useAuth() {
	return React.useContext(authContext);
}
