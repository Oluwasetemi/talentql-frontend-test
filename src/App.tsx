import * as React from 'react';
import {
	BrowserRouter as Router,
	Redirect,
	Route,
	Switch,
} from 'react-router-dom';
import { ToastContainer } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import styled from 'styled-components';
import { ProvideAuth } from './context/authContext';
import { useAuth } from './hooks/useAuth';
import Dashboard from './pages/Home';
import { Login } from './pages/Login';

function PrivateRoute({ children, ...rest }: any) {
	const auth = useAuth();

	return (
		<Route
			{...rest}
			render={({ location }) =>
				auth!.user ? (
					children
				) : (
					<Redirect
						to={{
							pathname: '/login',
							state: { from: location },
						}}
					/>
				)
			}
		/>
	);
}

const router: () => any = (): any => {
	return (
		<ProvideAuth>
			<Router>
				<Switch>
					<Route exact path="/login">
						<Login />
					</Route>
					<PrivateRoute exact path="/">
						<Dashboard />
					</PrivateRoute>
					<Redirect from="*" to="login" />
				</Switch>
			</Router>
		</ProvideAuth>
	);
};

export interface IColors {
	color: string;
	checked: boolean;
	isFixed: boolean;
	name: string;
}

export interface Filter {
	red: boolean;
	blue: boolean;
	green: boolean;
	'light-blue': boolean;
	gray: boolean;
}

export const AppWrapper = styled.div`
	width: 85%;
	margin: auto;
`;

function App(): React.ReactElement<any> {
	return (
		<>
			<ToastContainer
				position="top-right"
				autoClose={500}
				hideProgressBar={false}
				newestOnTop={false}
				closeOnClick
				rtl={false}
				pauseOnFocusLoss
				draggable
				pauseOnHover
			/>
			{router()}
		</>
	);
}

export default App;
