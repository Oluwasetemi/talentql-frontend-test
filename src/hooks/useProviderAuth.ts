import React from 'react';
import { fakeAuth } from '../App';

export function useProvideAuth() {
	const [user, setUser] = React.useState<string | null>(null);

	const signin = (cb: () => void) => {
		return fakeAuth.signin(() => {
			setUser('user');
			cb();
		});
	};

	const signout = (cb: () => void) => {
		return fakeAuth.signout(() => {
			setUser(null);
			cb();
		});
	};

	return {
		user,
		signin,
		signout,
	};
}
