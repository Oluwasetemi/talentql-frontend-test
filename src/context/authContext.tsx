import React from 'react';
import { useProvideAuth } from '../hooks/useProviderAuth';

export interface AppContextInterface {
	user: string | null;
	signin: (cb: () => void) => void;
	signout: (cb: () => void) => void;
}

export const authContext = React.createContext<AppContextInterface | null>(
	null,
);

export function ProvideAuth({ children }: { children: React.ReactNode }) {
	const auth = useProvideAuth();
	return <authContext.Provider value={auth}>{children}</authContext.Provider>;
}
