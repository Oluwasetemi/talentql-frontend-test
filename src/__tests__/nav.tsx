/* eslint-disable jest/no-disabled-tests */
import { act, fireEvent, render, screen } from '@testing-library/react';
import 'jest-styled-components';
import * as React from 'react';
import { Provider } from 'react-redux';
import wait from 'waait';
import App from '../App';
import { authContext } from '../context/authContext';
import { useProvideAuth } from '../hooks/useProviderAuth';
import store from '../store';

const mockUseLocationValue = {
	pathname: '/login',
	search: '',
	hash: '',
	state: null,
};

jest.mock('react-router-dom', () => ({
	...(jest.requireActual('react-router-dom') as {}),
	useLocation: jest.fn().mockImplementation(() => {
		return mockUseLocationValue;
	}),
}));

function NavApp() {
	const auth = useProvideAuth();
	return (
		<authContext.Provider value={auth}>
			<Provider store={store}>
				<App />
			</Provider>
		</authContext.Provider>
	);
}

test('nav component has a button log in', () => {
	const { getByTestId } = render(<NavApp />);
	const input = getByTestId(/login/i);
	expect(input).toMatchSnapshot();
});

test('fire button login in the has a button', async () => {
	const { findByTestId, getByTestId } = render(<NavApp />);
	const input = getByTestId(/login/i);

	act(() => {
		fireEvent.click(input);
	});

	const signOut = await findByTestId(/signout/i);
	expect(signOut).toBeInTheDocument();
});

test.skip('click on the logout button should log the fake user out', async () => {
	const { findByTestId } = render(<NavApp />);

	// const input = getByTestId(/login/i);
	screen.getByTestId(/login/i).click();
	let signOut = await findByTestId(/signout/i);

	act(() => {
		fireEvent.click(signOut);
		wait(200);
	});
	signOut = await findByTestId(/signout/i);
	// debug(signOut);
	// debug(container);

	expect(signOut).not.toBeInTheDocument();
});

test.skip('click on the logout button on the nav should log the fake user out', async () => {
	let res = render(<NavApp />);
	const { getByTestId, findByTestId } = res;
	// const { getByTestId, debug, container } = render(<NavApp />);
	const input = getByTestId(/login/i);
	await act(async () => {
		fireEvent.click(input);
		// await wait(200);
	});
	const signOut = await findByTestId(/logout/i);
	await act(async () => {
		fireEvent.click(signOut);
		await wait(200);
	});

	expect(signOut).toHaveTextContent('login');
});

test('click on the login button on the nav should log the fake user out', async () => {
	let res = render(<NavApp />);
	const { getByTestId, findByTestId } = res;
	// const { getByTestId, debug, container } = render(<NavApp />);
	const input = getByTestId(/signin/i);
	await act(async () => {
		fireEvent.click(input);
		// await wait(200);
	});
	const signOut = await findByTestId(/logout/i);
	expect(signOut).toHaveTextContent('logout');
});
