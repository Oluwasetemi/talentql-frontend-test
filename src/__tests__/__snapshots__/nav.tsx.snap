// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nav component has a button log in 1`] = `
<button
  data-testid="login"
>
  Log in
</button>
`;
